src/service.ts

```diff
@@ -331,6 +332,40 @@ export class Service {
     }
     return uri;
   }
+
+  static async downloadProject(project: Project, uri?: string) {
+    const zipFile: JSZip = new JSZip();
+    let zipName: string = "download.zip";
+    if (!isUndefined(uri)) {
+      zipName = `${uri}.zip`;
+    }
+    function addZipFile(file: File, folder?: any) {
+      if (file instanceof Directory) {
+        let folder: any;
+        if (!(file instanceof Project)) {
+          folder = zipFile.folder(file.name);
+          file.mapEachFile((file: File) => addZipFile(file, folder));
+        } else {
+          file.mapEachFile((file: File) => addZipFile(file, zipFile));
+        }
+      } else {
+        folder.file(file.name, file.data);
+      }
+    }
+    addZipFile(project);
```

Comment:
remove addZipFile recursion by creating a queue (below is a pseudo code) and make JSZip API comfomant:

```
   var queue = [];
   queue.push({filePrefix: '', file: project});
   while (queue.length > 0) {
      const { filePrefix, file } = queue.pop();
      const fileName = filePrefix + file.name;
      if (file instanceof Directory) {
         file.mapEachFile(f => queue.push({filePrefix: fileName + '/', file: f}));
         if (!(file instance Project)) zipFile.folder(fileName);
         continue;
      }
      zipFile.file(fileName, file.data);
   }
```

Reply:
const queue: Array<object> = [];
queue.push({filePrefix: "", file: project});
while (queue.length > 0) {
    const { filePrefix, file } = queue.pop(); //problem here
    ...
}

queue.pop is returning `ƒ pop() { [native code] }`
But queue becomes properly defined as: [{filePrefix:'', file: project}]  at L2

Reply:
> queue.pop is returning ƒ pop() { [native code] }

That's correct, `queue.pop`. What is returned by `queue.pop()` though? Actually, use `queue.shift()` instead.

Reply:
Correction: queue.pop() returns that function. Same for queue.shift().

Reply:
Really, is it how TS parses `{filePrefix: "", file: project}` ? Okay try `{filePrefix: "", file: project,}`

Reply:
Tried that, it didn't worked.

queue[0] returns `{filePrefix: "", file: project}`
queue.pop() does its job, it pops out.

But: const { filePrefix, file } = queue[0] doesn't work also. 
This is strange hmm hmm. There is some problem with data-structure.

Reply:
Worked for me when queue type is defined as `var queue: Array<{filePrefix: string; file: File}> = [];`

Reply:
Working code

```
    var queue: Array<{filePrefix: string; file: File}> = [];
    project.mapEachFile((f: File) => queue.push({filePrefix: '', file: f}));
    while (queue.length > 0) {
       const { filePrefix, file } = queue.shift();
       const fileName = filePrefix + file.name;
       if (file instanceof Directory) {
          file.mapEachFile((f: File) => queue.push({filePrefix: fileName + '/', file: f}));
          if (!(file instanceof Project)) zipFile.folder(fileName);
          continue;
       }
       zipFile.file(fileName, file.data);
    }
```